pylint/plugins/hass_return.py

```diff
@@ -0,0 +1,45 @@
+"""Plugin for return statements."""
+from __future__ import annotations
+
+from astroid import AsyncFunctionDef, Const, FunctionDef, Return
+from pylint.checkers import BaseChecker
+from pylint.interfaces import IAstroidChecker
+from pylint.lint import PyLinter
+
+
+class HassReturnFormatChecker(BaseChecker):  # type: ignore[misc]
+    """Checker for return statements."""
+
+    __implements__ = IAstroidChecker
+
+    name = "hass_return"
+    priority = -1
+    msgs = {
+        "W0016": (
+            "Remove explicit None in return",
+            "hass-return-none",
+            "Used when function returns nothing but has explicit 'return None'",
+        ),
+    }
+    options = ()
+
+    def visit_return(self, node: Return) -> None:
```

Comment:
Because I'm not that experienced with pylint/astroid to find undocumented `_get_return_nodes_skip_functions` :)
Thanks for the pointer!

I think pylint in many cases in used to force a good style rather than finding errors, so I don't see how this is different.

Regarding performance concerns I think it should be quite small and unnoticeable in a total run time. But I can try to measure the impact of extra plugin. Do you known what's the best way to see how long each individual check did take? Other than analyzing calls with cProfile.

Reply:
> Regarding performance concerns I think it should be quite small and unnoticeable in a total run time. But I can try to measure the impact of extra plugin. Do you known what's the best way to see how long each individual check did take? Other than analyzing calls with cProfile.

Performance profiling is a bit tricky, especially since at least some caching and multiprocessing is involved. I haven't tried it, but in theory you could write a decorator that sums the time all function calls to `visit_xxx` took. Just remember  to set `jobs=1` in `pyproject.toml` (for testing). I.e. something like this:
```py
from datetime import datetime, timedelta
from time import sleep
from typing import Callable

from typing_extensions import ParamSpec

P = ParamSpec("P")
runtime = timedelta(0)

def measure_time(func: Callable[P, None]) -> Callable[P, None]:
    def wrapper(*args: P.args, **kwargs: P.kwargs) -> None:
        start = datetime.now()
        func(*args, **kwargs)
        global runtime
        runtime += datetime.now() - start
    return wrapper


@measure_time
def visit_xxx():
    sleep(3)


if __name__ == "__main__":
    print(runtime)
    visit_xxx()
    print(runtime)
```

I haven't actually tested that for a pylint `visit_xxx` method yet, but it should work. You can print the result in the `HassReturnFormatChecker.close()` method.

Reply:
I did some tests and `_get_return_nodes_skip_functions` is about 60% slower than bottom-up implementation. This actually makes sense because `_get_return_nodes_skip_functions` iterates over all nodes in function subtree rather than just single path from return node to function root.

The plugin runtime is still very small, 120ms out of 10 minutes on my machine, which is ~0.02% of total runtime.